File: ai/app/quarter.py
"""분기 코드(YYYYQn) ↔ 정수 인덱스 변환.

인덱스 = year * 4 + (n - 1). 회귀 학습의 x축으로 사용.
"""


def to_index(q: str) -> int:
    if len(q) != 6 or q[4] != "Q" or q[5] not in "1234":
        raise ValueError(f"invalid quarter: {q!r}")
    year = int(q[:4])
    n = int(q[5])
    return year * 4 + (n - 1)


def from_index(idx: int) -> str:
    year, rem = divmod(idx, 4)
    return f"{year}Q{rem + 1}"


def next_quarter(q: str) -> str:
    return from_index(to_index(q) + 1)

File: ai-java/src/main/java/com/salesmap/ai/service/Quarter.java
package com.salesmap.ai.service;

/** 분기 코드(YYYYQn) ↔ 정수 인덱스. 인덱스 = year * 4 + (n - 1). */
public final class Quarter {

    private Quarter() {}

    public static int toIndex(String q) {
        if (q.length() != 6 || q.charAt(4) != 'Q' || "1234".indexOf(q.charAt(5)) < 0) {
            throw new IllegalArgumentException("invalid quarter: " + q);
        }
        int year = Integer.parseInt(q.substring(0, 4));
        int n = Character.digit(q.charAt(5), 10);
        return year * 4 + (n - 1);
    }

    public static String fromIndex(int idx) {
        int year = Math.floorDiv(idx, 4);
        int rem = Math.floorMod(idx, 4);
        return String.format("%04dQ%d", year, rem + 1);
    }

    public static String nextQuarter(String q) {
        return fromIndex(toIndex(q) + 1);
    }
}
